pkg/impersonation/impersonation.go

```diff
@@ -250,12 +250,24 @@ func (i *Impersonator) createRole(rules []rbacv1.PolicyRule) (*rbacv1.ClusterRol
 }
 
 func (i *Impersonator) rulesForUser() []rbacv1.PolicyRule {
-	rules := []rbacv1.PolicyRule{{
-		Verbs:         []string{"impersonate"},
-		APIGroups:     []string{""},
-		Resources:     []string{"users"},
-		ResourceNames: []string{i.user.GetUID()},
-	}}
+	rules := []rbacv1.PolicyRule{
+		{
+			Verbs:         []string{"impersonate"},
+			APIGroups:     []string{""},
+			Resources:     []string{"users"},
+			ResourceNames: []string{i.user.GetUID()},
+		},
+		{
+			Verbs:     []string{"impersonate"},
+			APIGroups: []string{"authentication.k8s.io"},
+			Resources: []string{"userextras/" + authcommon.ExtraRequestTokenID},
+		},
+		{
+			Verbs:     []string{"impersonate"},
+			APIGroups: []string{"authentication.k8s.io"},
+			Resources: []string{"userextras/" + authcommon.ExtraRequestHost},
+		},
```

Comment:
Note that the new extras are request-specific and therefore can't be bound to specific values (`ResourceNames`).

Reply:
I wasn't aware of this so maybe a dumb question but would it be possible to provide (and update) a list of token names that the user is owner of?

Reply:
Technically it's possible but tokens come and go. Do we want to constantly update the ClusterRole with the list?

Reply:
I don't know what you are planning to do with this value in the imperative API, but I think this might be a security issue. If we don't specify the `resourceNames` any user could change the value of the tokenID to be another token (even from a different user?) 

> Note that the new extras are request-specific and therefore can't be bound to specific values (ResourceNames).

If they are request-specific, can we add the token.Name from the request to the resourceNames? 

However, this wouldn't be possible when we create the impersonation resources when a PRTB or CRTB are created. I wouldn't add these extra permissions in this case.

Reply:
We could get the token from the request [here](https://github.com/rancher/rancher/blob/main/pkg/clusterrouter/proxy/proxy_server.go#L292)

Reply:
Impersonation authenticator now takes care of validating the token ownership.